### DISCUSSION TITLE
Optional route parameters

### DISCUSSION THREAD

**Discussion by jonasIv** [vercel/next.js#15526] [Help]
Hi 👋

I hope somebody can help me with the following problem.

### Context
I currently have this region-based page structure which I statically pre-render:

- `_app.js`
- `_document.js`
- `[regionId]`
  - `index.js`
  - `foo.js`
  - `bar.js`
  - ...

### Problem
In addition to what I have, I'd like to serve the same pages without a specified `regionId`, e.g. just `/foo`. Pages without a `regionId` in their path would use a default `regionId`. 
Example of resulting paths:
- `/foo` --> default region
- `/fr/foo` --> "fr" region
- `/de/foo` --> "de" region

The page has to be structured like this due to historical reasons. I have achieved this before by extending the path map. Now I would like to go the `getStaticPaths` way.

### Possible Solutions

- I have tried to make the route parameter optional, like `[[regionId]]`, but then I get this error: `Error: Optional route parameters are not yet supported ("[[regionId]]").`
- I could also use a "catch all" route like `[...slug].js` at the root level. It would require me to import all pages into one file and conditionally render the page based on the query object, right? Is NextJs smart enough to optimize the resulting pages, e.g. by eliminating the code of irrelevant page imports for each route?
- I could also re-create the page structure at the root level. But that doesn't seem to be an elegant solution.
- I might go back to extending the path map with region-based paths. But then I won't be able to use `getStaticProps` and `getStaticPaths`.

Any idea what the best solution is? Thanks for any help🙂

### Additional Notes
I'm considering to file an issue for optional route parameters. The error message makes it sound like they plan to add this feature but I couldn't find an open issue for it.

**Comment by SarKurd** [comment]
Next 9.5 just released, you should be be able to use optional catch all routes
https://nextjs.org/blog/next-9-5#optional-catch-all-routes

**Reply by onordgren** [reply]
Can this be used as a single optional route as well. Like [[regionId]] in the above example? I have a similar issue and a catch all route would still mean that all subpages needs to be handled by the same code instead of having dedicated pages.

**Reply by jonasIv** [reply]
@SarKurd, thanks for your answer. But I have a hard time understanding how optional catch all routes are helping here. Could you lay out the file structure?

**Reply by mebrein** [reply]
I created a new Next.js 9.5 project with only one file in the pages folder, named [[...slug.js]]. I think it's pretty cool!

``` javascript
import React from "react";
import { useRouter } from "next/router";

export default function Index(props) {
  const router = useRouter();

  const countRoute = router.query.slug?.length || 0;

  let regionName = "fr";
  let pageName = "";

  switch (countRoute) {
    case 0:
      // http://localhost:3000 -> pageName '', region 'fr', query: {}
      pageName = "";
      break;
    case 1:
      // http://localhost:3000/test -> pageName 'test', region 'fr'', query: {"slug":["test"]}
      pageName = router.query.slug[0];
      break;
    case 2:
      // http://localhost:3000/de/test -> pagename: 'test', region: 'de'', query: {"slug":["de","test"]}
      regionName = router.query.slug[0];
      pageName = router.query.slug[1];
      break;
  }

  return (
    <>
      <p>regionName: {regionName}</p>
      <p>pageName: {pageName}</p>
      <pre>query: {JSON.stringify(router.query, null, props)}</pre>
    </>
  );
}
```

**Reply by jonasIv** [reply]
@mebrein I see, interesting. Thanks for sharing. But then you basically circumvent the idea of the pages folder, right? In that case, I could also go back to customizing the path map.  I was hoping to get rid of manual routing logic by using dynamic routes.

**Reply by mebrein** [reply]
It's just a catchall route. Basically everything else in pages still works. Eg. if you create a test.js file inside pages. This file will be used instead of the [[...slug]].js.  From documentation, Predefined routes take precedence over dynamic route.

**Reply by jonasIv** [reply]
That's true, I can see your point. It's a solution I could work with, but it still feels like I'm duplicating routing logic. Importing `getStaticProps` and `getStaticPaths` from `test.js` into `[[...slug]].js` would probably help with keeping it lean.

**Reply by bambomax** [reply]
I've faced with the same issue, searched a little bit and found a solution. Since July 27th 2020, Nextjs supports rewrites: 
https://nextjs.org/blog/next-9-5#support-for-rewrites-redirects-and-headers 
https://nextjs.org/docs/api-reference/next.config.js/rewrites

next.config.js
```
module.exports = {
...,
async rewrites() {
    return [
      // check our routes
      {
        source: '/:slug*',
        destination: '/:slug*'
      },
      {
        source: '/:region',
        destination: '/'
      },
      {
        source: '/:region/:slug*',
        destination: '/:slug*'
      }
    ]
  }
}
```

and take the region name in the page component:
```
const Page = () => {
 const router = useRouter()
 console.log(router.query.region) // undefined for default region
 return {...}
}
```

```
https://....com/about // router.query.region = undefined
https://....com/fr/about // router.query.region = fr
```

**Reply by Phebonacci** [reply]
looks cool @bambomax but this won't work for SSG right?

**Reply by Phebonacci** [reply]
hi @jonasIv how do you extend the path map?

**Reply by bambomax** [reply]
> looks cool @bambomax but this won't work for SSG right?

Yes, I didn't find a better way :(

**Comment by mikestopcontinues** [comment]
What's blocking optional router parameters? It seems like they'd be easier to handle than optional catch-alls.

**Reply by Tylerian** [reply]
missing this aswell... any update?

**Reply by Vadorequest** [reply]
@timneutkens is it valid to use `[[optionalParam]]` in a route? It doesn't seem to work, while `[[...optionalParam]]` does.

**Reply by LarsEjaas** [reply]
Pump! I would also love this option to include "optional" folders like this!

**Reply by Akronae** [reply]
4 years later it appears that an optional single route parameter is too niche to have been implemented 👀
